NOnion/Network/TorGuard.fs

```diff
@@ -68,99 +68,147 @@ type TorGuard private (client: TcpClient, sslStream: SslStream) =
+                //FIXME: maybe continue instead of failing?
+                if command <> expectedCommandType then
+                    failwith (sprintf "Unexpected msg type %d" command)
+
+                use memStream = new MemoryStream (body)
+                use reader = new BinaryReader (memStream)
+                return Command.DeserializeCell reader expectedCommandType :?> 'T
         }
 
     member private self.ReceiveMessage () =
         async {
-            let! circuitId, command, body = self.ReceiveInternal ()
-            use memStream = new MemoryStream (body)
-            use reader = new BinaryReader (memStream)
-            return (circuitId, Command.DeserializeCell reader command)
+            let! maybeMessage = self.ReceiveInternal ()
+
+            match maybeMessage with
+            | Some (circuitId, command, body) ->
+                use memStream = new MemoryStream (body)
+                use reader = new BinaryReader (memStream)
+
+                return
+                    (circuitId, Command.DeserializeCell reader command) |> Some
+            | None -> return None
         }
 
     member private self.StartListening () =
         let listeningJob () =
             async {
-                //TODO: Handle socket closure and AsyncRead behaviour in case of socket being closed
-                while sslStream.CanRead do
-                    let! cid, cell = self.ReceiveMessage ()
-
-                    if cid = 0us then
-                        //TODO: handle control message?
-                        ()
-                    else
-                        match circuitsMap.TryFind cid with
-                        | Some circuit -> do! circuit.HandleIncomingCell cell
-                        | None -> failwith "Unknown circuit"
+                let rec readFromStream () =
+                    async {
+                        let! maybeCell = self.ReceiveMessage ()
+
+                        match maybeCell with
+                        | None -> ()
+                        | Some (cid, cell) ->
+                            if cid = 0us then
+                                //TODO: handle control message?
+                                ()
+                            else
+                                match circuitsMap.TryFind cid with
+                                | Some circuit ->
+                                    // Some circuit handlers send data, which by itself might try to use the stream
+                                    // after it's already disposed, so we need to be able to handle cancellation during cell handling as well
+                                    try
+                                        do! circuit.HandleIncomingCell cell
+                                    with
+                                        | :? ObjectDisposedException
+                                        | :? OperationCanceledException -> ()
+                                | None -> failwith "Unknown circuit"
```

Comment:
let's include cid in this error message

Reply:
Completely unrelated to PR, but addresed in a new commit